L1Trigger/TrackerTFP/src/DataFormats.cc

```diff
@@ -0,0 +1,752 @@
+    numUnusedBitsTracks_(+Process::end, TTBV::S),
+    numChannel_(+Process::end, 0)
+  {
+    setup_ = nullptr;
+    countFormats();
+    dataFormats_.reserve(numDataFormats_);
+    numStreams_.reserve(+Process::end);
+  }
+
+  // method to count number of unique data formats
+  template<Variable v = Variable::begin, Process p = Process::begin>
+  void DataFormats::countFormats() {
+    if constexpr(config_[+v][+p] == p)
+      numDataFormats_++;
+    if constexpr(++p != Process::end)
+      countFormats<v, ++p>();
+    else if constexpr(++v != Variable::end)
+      countFormats<++v>();
+  }
+
+  // proper constructor
+  DataFormats::DataFormats(const ParameterSet& iConfig, const Setup* setup) : DataFormats() {
+    iConfig_ = iConfig;
+    setup_ = setup;
+    fillDataFormats();
+    for (const Process p : Processes)
+      for (const Variable v : stubs_[+p])
+        numUnusedBitsStubs_[+p] -= formats_[+v][+p] ? formats_[+v][+p]->width() : 0;
+    for (const Process p : Processes)
+      for (const Variable v : tracks_[+p])
+        numUnusedBitsTracks_[+p] -= formats_[+v][+p] ? formats_[+v][+p]->width() : 0;
+    numChannel_[+Process::dtc] = setup_->numDTCsPerRegion();
+    numChannel_[+Process::pp] = setup_->numDTCsPerTFP();
+    numChannel_[+Process::gp] = setup_->numSectors();
+    numChannel_[+Process::ht] = setup_->htNumBinsQoverPt();
+    numChannel_[+Process::mht] = setup_->htNumBinsQoverPt();
+    numChannel_[+Process::sf] = setup_->htNumBinsQoverPt();
+    numChannel_[+Process::kfin] = setup_->kfNumWorker() * setup_->numLayers();
+    numChannel_[+Process::kf] = setup_->kfNumWorker();
+    transform(numChannel_.begin(), numChannel_.end(), back_inserter(numStreams_), [this](int channel){ return channel * setup_->numRegions(); });
+  }
+
+  // constructs data formats of all unique used variables and flavours
+  template<Variable v = Variable::begin, Process p = Process::begin>
+  void DataFormats::fillDataFormats() {
+    if constexpr(config_[+v][+p] == p) {
+      dataFormats_.emplace_back(Format<v, p>(iConfig_, setup_));
+      fillFormats<v, p>();
+    }
+    if constexpr(++p != Process::end)
+      fillDataFormats<v, ++p>();
+    else if constexpr(++v != Variable::end)
+      fillDataFormats<++v>();
+  }
+
+  // helper (loop) data formats of all unique used variables and flavours
+  template<Variable v, Process p, Process it = Process::begin>
+  void DataFormats::fillFormats() {
+    if (config_[+v][+it] == p) {
+      formats_[+v][+it] = &dataFormats_.back();
+    }
+    if constexpr(++it != Process::end)
+      fillFormats<v, p, ++it>();
+  }
+
+  // converts bits to Stub
+  template<typename ...Ts>
+  void DataFormats::convertStub(const TTDTC::BV& bv, tuple<Ts...>& data, Process p) const {
+    TTBV ttBV(bv);
+    extractStub(ttBV, data, p);
+  }
+
+  // helper (loop) to convert bits to Stub
+  template<int it = 0, typename ...Ts>
+  void DataFormats::extractStub(TTBV& ttBV, std::tuple<Ts...>& data, Process p) const {
```

Comment:
1) General comment -- Much of the code in this class is extremely difficult to understand. Each line of C++ is often using several layers of function calls, and lots of rarely used C++ features. To ensure that the code can be maintained, I encourage you to keep things simple where possible. 
2) To help comprehension, we need more precision with the comments and the variable names. I believe extractStub() doesn't "convert bits to stub", but rather switches between two different representations of a stub: all variables packed in single bit vector to nutple of digitized variables. "data" gives no clue as to what it represents. And as this function  has non-const ref. arguments, it's not obvious which are the input and which are the output. Best to put input arguments first. And perhaps add "Out" to the name if still not clear.

Reply:
done

Reply:
Please check again this comment, which is general, and applies to many functions in this class. Also, point (2) has not yet been fully addressed, even for the extractStub() function. e.g. The variable name "data" has not yet been changed to give a clearer idea of what it represents, or whether it is an input or an output argument or both.